server/src/main/java/org/elasticsearch/index/TimestampBounds.java

```diff
@@ -18,10 +20,12 @@ public class TimestampBounds {
     private final long startTime;
     private volatile long endTime;
 
-    TimestampBounds(IndexScopedSettings scopedSettings) {
-        startTime = scopedSettings.get(IndexSettings.TIME_SERIES_START_TIME).toEpochMilli();
-        endTime = scopedSettings.get(IndexSettings.TIME_SERIES_END_TIME).toEpochMilli();
-        scopedSettings.addSettingsUpdateConsumer(IndexSettings.TIME_SERIES_END_TIME, this::updateEndTime);
+    TimestampBounds(IndexMetadata indexMetadata, @Nullable IndexScopedSettings scopedSettings) {
+        startTime = indexMetadata.getTimeSeriesStart().toEpochMilli();
+        endTime = indexMetadata.getTimeSeriesEnd().toEpochMilli();
+        if (scopedSettings != null) {
+            scopedSettings.addSettingsUpdateConsumer(IndexSettings.TIME_SERIES_END_TIME, this::updateEndTime);
+        }
```

Comment:
Ah! I see. What a pain.

I wonder if it's better to have the bounds object be immutable and to move the replacement of the end time to the `IndexSettings`. Or wrap this in something mutable over there when in tsdb mode. I'd wanted to avoid having a `volatile` read from `IndexSettings` for every document write which is why I made this mutable. The way it is now you only do the volatile read when you are in tsdb mode.

The mutability here is annoying. Sorry!

Reply:
I tried to the following: ba5460c10e0cd1dfcaa654ebe4a6ada8eac12bfb
Which updates the timestampBounds in `IndexSettings` if end_time setting changes and makes `TimestampBounds` immutable.